idaes/generic_models/unit_models/surrogate_model.py

```diff
@@ -0,0 +1,135 @@
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia
+# University Research Corporation, et al. All rights reserved.
+#
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+"""
+Generic template for a surrogate unit model.
+"""
+
+from pyomo.environ import Reference, SolverFactory
+from pyomo.network import Port
+from pyomo.common.config import ConfigValue, In
+
+from idaes.core.process_base import declare_process_block_class, \
+    ProcessBlockData
+from idaes.core.util.exceptions import ConfigurationError
+from idaes.core.util.model_statistics import degrees_of_freedom
+import idaes.logger as idaeslog
+
+__author__ = "Jaffer Ghouse"
+
+# Set up logger
+_log = idaeslog.getLogger(__name__)
+
+
+@declare_process_block_class("SurrogateModel")
+class SurrogateModelData(ProcessBlockData):
+    """
+    This is the class for a surrogate unit model. This will provide a user
+    with a generic skeleton to add a surrogate unit model.
+    """
+
+    # Create Class ConfigBlock
+    CONFIG = ProcessBlockData.CONFIG()
+    CONFIG.declare("dynamic", ConfigValue(
+        default=False,
+        domain=In([False]),
+        description="Dynamic model flag",
+        doc="""Surrogate model assumed to be steady-state."""))
+
+    def build(self):
+        """
+        General build method for SurrogateModelBlockData.
+
+        Inheriting models should call `super().build`.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        super(SurrogateModelData, self).build()
+
+    def add_ports(self, name=None, member_list=None, doc=None):
+        """
+        This is a method to build Port objects in a surrogate model and
+        populate this with appropriate port members as specified. User can add
+        as many inlet and outlet ports as required.
+
+        Keyword Args:
+            name : name to use for Port object.
+            dict : dictionary containing variables to be added to the port
+            doc : doc string for Port object
+
+        Returns:
+            A Pyomo Port object and associated components.
+        """
+        # Validate that member_list is a dict
+        if not isinstance(member_list, dict):
+            raise ConfigurationError(
+                "member_list should be a dictionary "
+                "with the keys being the name assigned(strings) and "
+                "values being the variable objects declared in the model.")
+
+        # Create empty Port
+        p = Port(noruleinit=True, doc=doc)
+        setattr(self, name, p)
+
+        # Create References for port members
+        for k in member_list.keys():
```

Comment:
This will depend a bit on how the surrogate itself is built, but do we need to make `References` here?

This was necessary in the core models to make it look lie the state variables were directly indexed by time (instead of being on a `Block` indexed by time). Here I would have guessed that any variable would be directly indexed by time (as surrogates probably lack the block-hierarchical structure of other models). If that is the case, then we should skip the `References`.

Reply:
An important thing to remember here is that `Port` members need to be indexed by time in order to link up with other IDAES models.

That means that either the internal surrogate model variables need to be indexed by time, or we need to add a time index to them in the `Ports` (via `References`). We need to decide which approach we are going to use and clearly document it.

Reply:
@andrewlee94 Thanks for this comment. I am leaning towards model variables indexed by time and leaving it to user. See test for an example. I have also documented this.

Reply:
@jghouse88 I think indexing the variables (and surrogate constraints) by time is the best approach as well, and the draft surrogate API supports this. This makes it a lot easier at the `SkeletonModel` level as you can just add variables directly to `Ports` as required.